File: TorchMobile/app/src/main/java/magshimim/torchmobile/networking/NetworkManager.java
package magshimim.torchmobile.networking;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;

import magshimim.torchmobile.networking.protos.TorchMessageOuterClass;
import magshimim.torchmobile.utils.Event;

public class NetworkManager {
    private boolean working;
    private int port;
    private final Object writeLock;
    private String host;
    private DataInputStream inputStream;
    private DataOutputStream outputStream;
    private Socket socket;
    private Thread workerThread;

    public Event<Void> onConnected;
    public Event<Boolean> onDisconnected;
    public Event<Exception> onError;
    public Event<TorchMessageOuterClass.TorchMessage> onMessage;

    public NetworkManager() {
        onConnected = new Event<>();
        onDisconnected = new Event<>();
        onError = new Event<>();
        onMessage = new Event<>();
        working = false;
        writeLock = new Object();
    }

    public boolean isConnected() {
        return working &&
                socket != null &&
                socket.isConnected();
    }

    public void connect(String host, int port) {
        if(working || workerThread != null)
            throw new IllegalStateException("Connection has already been established or started");

        this.host = host;
        this.port = port;
        workerThread = new Thread(new Runnable() {
            @Override
            public void run() {
                receiveLooper();
            }
        });
        workerThread.start();
    }

    public void close() {
        close(false);
    }

    public synchronized void send(TorchMessageOuterClass.TorchMessage message) {
        if(message == null)
            return;

        if(!canWrite())
            throw new IllegalStateException("Output stream is closed");

        try {
            synchronized (writeLock) {
                message.writeDelimitedTo(outputStream);
            }
        } catch (IOException e) {
            if(working) {
                onError.invoke(e);
                close(true);
            } else e.printStackTrace();
        }
    }

    private boolean canRead() {
        return isConnected() &&
                !socket.isInputShutdown() &&
                inputStream != null;
    }

    private boolean canWrite() {
        return isConnected() &&
                !socket.isOutputShutdown() &&
                outputStream != null;
    }

    private void close(boolean hadError) {
        if(!working)
            return;

        working = false;

        if(workerThread != null) {
            if(workerThread.getId() != Thread.currentThread().getId() && workerThread.isAlive())
                workerThread.interrupt();
            workerThread = null;
        }

        if(inputStream != null) {
            try {
                inputStream.close();
            } catch (IOException e) {
                hadError = true;
                e.printStackTrace();
            }
            inputStream = null;
        }

        if(outputStream != null) {
            try {
                outputStream.close();
            } catch (IOException e) {
                hadError = true;
                e.printStackTrace();
            }
            outputStream = null;
        }

        if(socket != null) {
            if (socket.isConnected())
                try {
                    socket.close();
                } catch (IOException e) {
                    hadError = true;
                    e.printStackTrace();
                }
            socket = null;
        }

        onDisconnected.invoke(hadError);
    }

    private void receiveLooper() {
        boolean hadError = false;

        try {
            socket = new Socket(host, port);
            inputStream = new DataInputStream(socket.getInputStream());
            outputStream = new DataOutputStream((socket.getOutputStream()));
            working = true;
            onConnected.invoke(null);

            while (canRead()) {
                TorchMessageOuterClass.TorchMessage message = TorchMessageOuterClass.TorchMessage
                        .parseDelimitedFrom(inputStream);
                onMessage.invoke(message);
            }
        } catch (IOException e) {
            hadError = true;
            if(working) onError.invoke(e);
            else e.printStackTrace();
        }

        close(hadError);
    }
}

File: TorchDesktop/TorchDesktop/Networking/NetworkManager.cs
using Google.Protobuf;
using System;
using System.Diagnostics;
using System.Net.Sockets;
using System.Threading;
using TorchDesktop.Networking.Protos;

namespace TorchDesktop.Networking
{
    public delegate void Callback();
    public delegate void Callback<T>(T value);

    public class NetworkManager
    {
        private bool working;
        private int port;
        private string host;
        private NetworkStream stream;
        private TcpClient client;
        private Thread workerThread;

        public bool Working
        {
            get { return working; }
        }

        public event Callback Connected;
        public event Callback<TorchMessage> Message;
        public event Callback<string> Error;
        public event Callback<bool> Disconnected;

        public NetworkManager()
        {
            working = false;
        }

        public void Connect(string host, int port)
        {
            if (working || workerThread != null)
                throw new InvalidOperationException("Already working");

            this.host = host;
            this.port = port;
            workerThread = new Thread(ReceiveLooper);
            workerThread.Start();
        }
        
        public void Close()
        {
            working = false;
            if (workerThread.IsAlive)
                workerThread.Abort();
            if (client.Connected)
                client.Close();
        }

        public void Send(TorchMessage message)
        {
            if (!working || stream == null || !stream.CanWrite)
                throw new InvalidOperationException("Cannot write message");
            
            message.WriteDelimitedTo(stream);
        }

        private void CloseConnection(bool hadError = false)
        {
            working = false;
            client.Close();
            Emit(nameof(Disconnected), hadError);
        }

        private void ReceiveLooper()
        {
            bool error = false;

            try
            {
                client = new TcpClient(host, port);
                stream = client.GetStream();
                working = true;
                Emit(nameof(Connected));

                while (working && client.Connected)
                {
                    TorchMessage message = TorchMessage.Parser.ParseDelimitedFrom(stream);
                    Emit(nameof(Message), message);
                }
            }
            catch (Exception e)
            {
                error = true;
                Emit(nameof(Error), "Disconnected");
                Emit(nameof(Disconnected), true);
            }

            if (working)
                CloseConnection(error);
        }

        private void Emit(string eventName, object parameter = null)
        {
            switch (eventName)
            {
                case nameof(Connected):
                    Connected?.Invoke();
                    break;
                case nameof(Message):
                    TorchMessage message = parameter as TorchMessage;
                    if (message != null)
                        Message?.Invoke(message);
                    else
                        Trace.TraceWarning("Tried to emit empty message");
                    break;
                case nameof(Error):
                    string error = parameter as string;
                    if (error != null)
                        Error?.Invoke(error);
                    else
                        Trace.TraceWarning("Tried to emit empty error");
                    break;
                case nameof(Disconnected):
                    if (parameter?.GetType() == typeof(bool))
                        Disconnected?.Invoke((bool)parameter);
                    else
                        Trace.TraceWarning("Tried to emit non boolean");
                    break;
                default:
                    Trace.TraceWarning($"Unknown event name '{eventName}'");
                    break;
            }
        }
    }
}
